### DISCUSSION TITLE
Notes on porting a manuscript from Scrivener

### DISCUSSION THREAD

**Discussion by instancezero** [vkbo/novelWriter#1793] [General]
I'm moving my WIP from Scrivener to NW with this process:
- Select a scene in Scrivener
- Copy it to LibreOffice Writer, save the document
- Use pandoc to convert the ODT to MD
- Run a little script to tweak the pandoc output and convert it to NW's dialect of MD
- Create a scene in NW
- Open the scene and copy the scene header
- Import the MD text
- Paste the scene header back in
- Copy comments from Scrivener and add as % comments near the paragraph they refer to

My first suggestion is a simple one: instead of the "this will overwrite yes/no" dialog, present a dialog that offers Replace/Append/Cancel as options, thus eliminating the copy/paste of the scene header.

It would also be helpful in the file selection dialog defaulted to all file formats (.txt, .md, and .nwd) with options for a specific format in the drop-down.

I miss the ability to highlight and annotate text in Scrivener. Actually, the requirement to highlight text is a little irritating; often I just want to add a notation at a specific point. My primary use case is to quickly add some feedback from a critique group in real time. Would you consider adding an inline comment capability? I'd suggest:

```[note]some MD free note here[/note]```

Although I'm sure that others value the ability to attach the comment to some text. Maybe something like

```Jane found it peculiar [note/ over-use of peculiar here] that...```

for a stand-alone comment, and

```[note "her" is ambiguous here]he thought it was hers[/note]```

for a comment that applies to a segment of text. No MD allowed in the comment itself, of course.

As a bonus it would be cool if those notes got exported as comments in an ODT file.

**Comment by vkbo** [comment]
> I'm moving my WIP from Scrivener to NW with this process:
> [...]
> My first suggestion is a simple one: instead of the "this will overwrite yes/no" dialog, present a dialog that offers Replace/Append/Cancel as options, thus eliminating the copy/paste of the scene header.

The current import feature isn't well suited for repeated imports, so I tend to prefer to do it in bulk.

novelWriter has a split document feature that can split by any header level you want. I have used this to import large public domain texts that I use for stress testing novelWriter. Including the full King James bible, Moby Dick and A Tale of Two Cites. It is still a bit of work, but it is quite manageable. It's a lot easier than to import many small clips.

Anyway, Replace/Append/Cancel should be doable. It is fairly straightforward.

> It would also be helpful in the file selection dialog defaulted to all file formats (.txt, .md, and .nwd) with options for a specific format in the drop-down.

Sure, that can be managed.

> I miss the ability to highlight and annotate text in Scrivener. Actually, the requirement to highlight text is a little irritating; often I just want to add a notation at a specific point. My primary use case is to quickly add some feedback from a critique group in real time. Would you consider adding an inline comment capability?

There is already a feature idea for a general note comment (not inline) here: #1133.

Highlighting has been implemented (#705, #1715) and will be available in the 2.4 release (currently in beta). However, highlighting doesn't come with a note capability.

I would very much like to add such a feature, but since this is not rich text, it is a little tricky to allow free form text inside tags and parse them reliably with just simple regex. I haven't come up with a good implementation idea yet, but I'm very open to suggestions.

Perhaps we could add them in a similar way to footnotes (#342). The idea there is to add a marker to put in the text, and a way to specify the text that goes into the footnote as a separate text block. Perhaps something similar to special comments, so we can avoid adding more ways to format additional text.

The idea is to let a `[fn:<ref>]` or `[footnote:<ref>]` tag indicate where the footnote is inserted, and the `<ref>` points to the corresponding text. The text can be placed anywhere, so the logical placement would be next to the paragraph where it is defined. This is close enough to be easy to write and edit, and doesn't matter for exports since the text is indexed in advance, so any such reference can be known before tokenizing the text.

My immediate thought is that an inline note feature can easily be implemented in a similar way with `[note:<ref>]`.

Outputting each of them to ODT would be easy enough to manage.

How does this sound?

**Reply by instancezero** [reply]
I think that would work well. In every editor I've seen that supports comments, They're always in some sort of an aside as it is, so this is reasonably congruent with what most people would expect. This also avoids all sorts of potential markup nesting issues and the corresponding increase in regex complexity.  All your approach is doing is asking for users to come up with values for `<ref>` and since notes are by definition not going to wind up in the final manuscript not much thought needs to go into that.

After reading the referenced issues, I'm glad you're sticking with Markdown. Granted it's more cumbersome than working with a rich text editor, but it makes it much easier to move forward with the markup outlined in #1784 knowing I can write a stand-alone tool to pull the data and move forward. If the data were stored in a more complex format that would be far more difficult.

**Reply by vkbo** [reply]
I have considered adding a JSON block in the nwd files for storing meta data and allowing the user to add additional meta data in a more rich text like fashion. I'm uncertain if that is wise. I find the note feature in most office applications to be clunky and sometimes unwieldy.

In any case, I would definitely add a button-click approach to insert them so that the user doesn't have to remember any syntax. For instance, for footnote. Say I have the following text paragraph:

```md
I want to insert a footnote at the end of this sentence at the cursor.| And the text continues.
```

Then, with the cursor at `|`, the user creates a footnote, and it changes to:

```md
I want to insert a footnote at the end of this sentence at the cursor.[fn:f3g5] And the text continues.

%Footnote.f3g5: |
```

Where `f3g5` is just some auto-generated unique handle and the footnote text block is inserted after the current paragraph and the cursor moves there automatically.

Likewise, an inline note on a selected text could automatically generate:

```md
I want to [note:nh74]insert a footnote[/note] at the end of this sentence at the cursor.[fn:f3g5] And the text continues.

%Note.nh74: |

%Footnote.f3g5: 
```

This will interfere with my other idea for note blocks, so either this should use a different word than "Note", or the note block should.

**Reply by vkbo** [reply]
The benefit of this approach is that the infrastructure for `%Story.term` can be repurposed as-is in the index parser. So this is a very flexible framework, and should be easy to maintain and extend. The shortcode format is also already fairly easy to parse with regex, and I already have most of the code in place for that. It is currently used for the `[vspace]` functionality to insert vertical space in the text.

I've also made the regex processing a separate function. Adding some of the mentioned functionality would benefit from comments allowing in-line formatting as well, which can be achieved. The current formatting parser is located here:

https://github.com/vkbo/novelWriter/blob/2480c7ae9f02b453599aa047a3aed445bfa24f6e/novelwriter/core/tokenizer.py#L964-L999

**Reply by instancezero** [reply]
Alternate words could be "Comment" or "Rem" (for remark). I only used "note" because it's shorter than "Comment". If you don't foresee a use for it, maybe that could be shortened to just "c", or consistent with footnote, "cn" for "comment note". 

I suspect that the `%command.term: text (possibly following some command-specific syntax)` approach will supply all the room that's needed for metadata for some time to come.

Slightly tangential: you can avoid the problem with @time referring to timeline rather than time-of-the-scene by storing scene time as `%story.time: text` and leaving @time as-is.

**Reply by vkbo** [reply]
I do have an idea for how to handle it with `@time` as well. I added a display name feature for tags a while back, so the user could add their own numerical code for the time tag and add a display name that can be used in a timeline overview. Like so:

```md
### Scene

@time: Day1
```

And in the notes:

```md
# Timeline

## Day 1

@tag: Day1 | Thursday April 4 2024

## Day 2

@tag: Day2 | Friday April 5 2024
```

But it is a little clunky in that each entry needs a heading and a `@tag` definition.

An alternate implementation could be simply:

```md
### Scene

%Time.Day1: Thursday April 4, 2024
```

And the information collected like other special comments. No need to define them anywhere else.

The benefit of either approach is that the time key can be implemented in a sortable manner as the users see fit, and the display text is completely free form and would allow for inventing any calendar you'd like.

**Reply by instancezero** [reply]
Yes, attempting to standardize time is futile. Everything in my WIP is relative to the opening chapter, mostly in days, so +3d is three days later and -6mo is six months earlier, and t=0 is never explicitly set on a calendar. I was thinking of the person who wanted bidirectional integration with a timeline program. In that case it seems a specific `%Time...` construct for that purpose and a standardized time format would be the most viable approach.

I'll continue with `%story.time:` until something else is available in NW, then write a quick script to transform my notation into that of the final implementation. More joys of Markdown!

**Reply by vkbo** [reply]
My own project has a custom calendar, so standardising it is just futile. Also, having notation based on English is also limiting. It's already enough that the keywords are English. The user should be free to use whatever language they want for the actual content.

Anyway, thanks for this and the other discussion. They have helped me figure out a way to implement a number of features in a flexible way. I really want to make sure novelWriter doesn't have too many variations of meta data syntax. I'm quite happy with what we came up with for `%Story` and how we can extend it to multiple features.

I moved the footnote ticket out of the 2.4 milestone and into 2.5 as 2.4 was getting rather large, but I think I will start implementing the needed framework in 2.5 using the discussed implementation here. It should be fairly easy to tag on the note, comment and story features. Maybe I can bundle them all into one release or maybe split them between 2.5 and 2.6 to not overload the user and also get a few other things in.

Most of the early 2.x releases have been huge, and taken a very long time to complete. I do want to shorten the release cycle a little.

**Reply by instancezero** [reply]
I was thinking for the import/export use case, the "pre-set format" would be something like ISO8601, which is also why I'd make it a specific use bit of metadata. Those of us with custom calendars could keep doing what we're doing and ignore that use case.

As a user I have no problem with your release cycle. I'm used to projects where point releases are a year apart or more.

As a developer, especially with a) my more rudimentary Python skills and b) resisting my inclination to think "these tokens should be refactored as a dict", it's hard to contribute because you're actively doing a lot of work on the code and have already overhauled tokenizer.py.

Just yesterday I realized I should stop trying to muck with the code and get back to my [curse words redacted] manuscript. I wrote a [blog post](https://alanlangford.com/2024/04/03/the-power-and-perils-of-knowing-how-to-program/) about getting suckered into coding holes. Now I'm happy to be leaving my contributions at the conceptual level!

**Reply by vkbo** [reply]
> I was thinking for the import/export use case, the "pre-set format" would be something like ISO8601, which is also why I'd make it a specific use bit of metadata. Those of us with custom calendars could keep doing what we're doing and ignore that use case.

In my social media profile I have "ISO 8601 Enthusiast" listed, so I am definitely a fan!

However, since I mostly write SciFi, I don't only deal with calendars tailored to the orbital properties of Earth. Hence I don't want to lock down format. I think letting the user chose between a few sort functions could solve most of it. Like alphabetical, and positive/negative numerical, etc. ISO8601 is sortable by plain alphabetical/numerical sorting, so it's a good format and would still work.

> As a user I have no problem with your release cycle. I'm used to projects where point releases are a year apart or more.

novelWriter is still in the feature-adding part of its life cycle. The main reason I manage to retain a good level of stability is that I currently have 99.51% test coverage, with multiple overlapping tests. It lets me know immediately if I break anything. There are certain things not covered, like some GUI features like drag and drop. These are notoriously difficult to test with the Qt framework, and is quite buggy because Qt too don't test it properly. I also use high quality linting tools to assist me, which tend to catch type inconsistencies before they become bugs. I had a lot more type-related bugs early on.

> As a developer, especially with a) my more rudimentary Python skills and b) resisting my inclination to think "these tokens should be refactored as a dict", it's hard to contribute because you're actively doing a lot of work on the code and have already overhauled tokenizer.py.

Heh. When I first started this project, Python dicts weren't ordered, so that wasn't possible. Unless you meant the individual tokens in the list. Using the tuple is more efficient, but less readable and a little more error prone. Although the type checking tool lets me know if I make a mistake.

The project structure is also still stored in memory as a list preserving the order, and a dict for the data. The double storage is too closely tied to many features that I can't be bothered rewriting it, and it has worked for years and is pretty solid.

I am constantly updating the code when Python changes, and when I learn new things about how to optimally use Python. So this project is keeping me more on my toes than what I do at my day job, which is also mostly writing Python code.

> Just yesterday I realized I should stop trying to muck with the code and get back to my [curse words redacted] manuscript. I wrote a [blog post](https://alanlangford.com/2024/04/03/the-power-and-perils-of-knowing-how-to-program/) about getting suckered into coding holes. Now I'm happy to be leaving my contributions at the conceptual level!

I read that earlier today, yeah. I too need to step away from the code some times to do some actual writing!

**Reply by instancezero** [reply]
I think I was ranting just last week about SF where humans land on a brand new planet and the occupants talk about events that were "x years ago" and I'm screaming "how long is one of their years, anyway?" Not to mention that if humans got there at near-light speed their own timeline is completely different from Earths anyway. Relativistic physics is one big reason why my work is set in the near future on Earth!

I'm not aware of timeline tool that accounts for time-space dilation so I'm assuming that people who want to use a third party tool are going to be forced into an Earth calendar. Even then ISO8601 can accommodate "Stardate 5503" and crowbar in a fractional part. My guess is that the main application would be for something like historical fiction where nailing the dates would be important for a credible story.

**Reply by vkbo** [reply]
I'm a physicist by education, so Einstein doesn't scare me. 😄 

One of my favourite authors, Alastair Reynolds, who's also a physicist, handles near light speed and relativistic time tracking quite well and makes it a part of the plot. I like his approach, although I don't need to do that in any of my current stories.

Anyway, I mostly track time in my writing for my own sake, for internal consistency.

**Reply by instancezero** [reply]
My background is physics too. No fear but lots of respect. ;) I'll look up the work of Mr. Reynolds.
